Extend dataset dependencies

Lee-W opened the pull request:
Related: https://github.com/apache/airflow/issues/40039
Depend on: https://github.com/apache/airflow/pull/40693

## What
In https://github.com/apache/airflow/pull/40693, we introduce a many-to-many relationship between dataset and dataset alias. This PR allows `GET /object/dataset_dependencies` to retrieve the dataset aliases linked to that dataset as well.

The following is example graph

![image](https://github.com/user-attachments/assets/cebc70b5-2766-40d7-86fb-09e0e844ca18)



<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Lee-W committed:
feat(www/views.py): add dataset alias node to dataset_dependencies

Lee-W committed:
feat(www/view): extend dataset-alias to connect with its resolved datasets

Lee-W committed:
refactor: rename expand_as_end_nodes as expand_as_dep_nodes

Lee-W committed:
refactor(serialization): extract DagDependency as a separate module

Lee-W committed:
refactor(dataset): refactor expand_as_dag_nodes (only the target part)

Lee-W committed:
refactor(dataset): rewrite the source part of expand_as_dag_nodes

Lee-W committed:
refactor(datasets): rename expand_as_dag_nodes as iter_dag_deps

Lee-W committed:
docs(datasets): add docstring for iter_dag_deps

Lee-W committed:
refactor(www): refactor dataset_dependencies logic

Lee-W committed:
style: fix mypy issue

bbovenzi reviewed (approved):
the webserver part looks good

uranusjr commented on airflow/datasets/__init__.py:
```diff
@@ -185,6 +186,14 @@ def evaluate(self, statuses: dict[str, bool]) -> bool:
     def iter_datasets(self) -> Iterator[tuple[str, Dataset]]:
         raise NotImplementedError
 
+    def iter_dag_deps(self, *, source: str, target: str) -> Iterator[DagDependency]:
```
Any concerns just naming this `iter_dag_dependencies` to match the return value type name?

Lee-W replied:
Nope, `iter_dag_dependencies` is good. Will update it

uranusjr commented on airflow/datasets/__init__.py:
```diff
@@ -271,6 +306,20 @@ def iter_datasets(self) -> Iterator[tuple[str, Dataset]]:
+
+        :meta private:
+        """
+        dag_deps: set[DagDependency] = set()
```
Is the order of DAG dependencies significant? If not, I think we can deduplicate at where this function is called instead. If we have aliases nested in conditions (or just nested conditions), each nesting would have a set doing deduplication, which isn’t efficient IMO.

Or, do we need to deduplicate dependencies at all…?

Lee-W replied:
> Is the order of DAG dependencies significant? If not, I think we can deduplicate at where this function is called instead. 

I think not. but I'll test out.

> Or, do we need to deduplicate dependencies at all…?

I would probably a +0? e.g., `scheudle=[Datset("test-1"), Dataset("test-1")]` should we generate 2 `Dataset("test-1")`? On the other hand, it's also what users want 🤔  I feel it's more like a UI thing and would like to know how @bbovenzi  think

Lee-W replied:
oh, in the current dateset_dependecnies endpoint, I think it'll be filtered out anyway. so we can also defer the deduplication later as well

Lee-W replied:
I feel that not deduplicating might make it even more complicated.

uranusjr replied:
Can you elaborate?

Lee-W replied:
Sure.

currently, a DAG with `schedule=[Dataset("1"), Dataset("1")]` will generate the following graph.


```mermaid
flowchart LR
    dataset1 --> DAG1
```

By not deduplicate, my understanding is we want to have something like this. Then we'll have to mark that the frist `Dataset("1")` and the second `Dataset("1")` are different. It's not undoable. Just not sure whether it makes sense.

```mermaid
flowchart LR
    dataset1 --> DAG1
    dataset1 --> DAG1
```

Lee-W replied:
AFAIK, we do not have `DatasetAll` and `DatasetAny` in the graph. so `schedule=Dataset("1") & Dataset("1") | Dataset("1")` would be something similar to the above comment.

uranusjr replied:
In this particular situation, I wonder if we should just deduplicate in DatasetAll and DatasetAny. I anticipate there might be other places we should dedup, but at least for those there’s a simple math solution.

Lee-W replied:
What I'm thinking now is whether there's a situation we really don't want deduplication. Or is this more like a performance issue?

uranusjr replied:
For DatasetAll and DatasetAny specifically, I don’t think there’s a situation we need to keep the original user input. I’d even argue it’s an oversight I did not implement dedup in `__and__` and `__or__` in the initial release.

There are cases where we can’t dedup in all/any but need further cleanup in the UI though, say `("a" | "b") & ("a" | "c")`. This cannot be simplified, but we only want one line to `a` instead of two. But that’s a UI consideration that should only be done either in the API layer or frontend (React).

Lee-W replied:
> say ("a" | "b") & ("a" | "c"). This cannot be simplified, but we only want one line to a instead of two. 

I guess I'm a bit confused. What do you expect the UI to look like in this situation? It would look something like the following now.

```mermaid
flowchart LR
    a --> DAG
    b --> DAG
    c --> DAG
```

If we don't duplicate (for cases that can't actually be dedup), I would probably want something like this. 

```mermaid
flowchart LR
    a --> Or1
    b --> Or1
    a --> Or2
    c --> Or2

    Or1 --> And
    Or2 --> And

    And --> DAG
```

Lee-W replied:
*not related to the whether to deduplicate discussion*

I've moved the deduplication out of `iter_dag_dependencies`

uranusjr replied:
By “now” above you mean before this PR right? That’s what I would expect the graph to look like. So we still need some dedup logic when rendering the UI, regardless what we do in all/any. But I want those to be strictly in the UI layer (i.e. somewhere in API or frontend), not `iter_dag_dependencies`.

Lee-W replied:
Got your point! Yep, it's no longer in `iter_dag_dependencies`. But it still happens in `airflow/serialization/serialized_objects.py`. So I think it's still not ` strictly in the UI layer (i.e. somewhere in API or frontend),` But in the "current" (before this PR) version, we're using `iter_datasets` to [detect_dag_dependencies](https://github.com/apache/airflow/blob/9ec9eb79a0cc845d86e7380c73269d2ee1d3c210/airflow/serialization/serialized_objects.py#L963). ~~On the other hand, the [outlet part](https://github.com/apache/airflow/blob/9ec9eb79a0cc845d86e7380c73269d2ee1d3c210/airflow/serialization/serialized_objects.py#L946-L955) did not deduplicate.~~ I was wrong, [outlet](https://github.com/apache/airflow/blob/0eaa6117bf0310be798ad2be5c82905bd6fa30b5/airflow/serialization/serialized_objects.py#L1362) was also de-duplicated

I think what you suggest is better. (let the API layer do the deduplication job, I'll probably do it [here](https://github.com/apache/airflow/blob/9ec9eb79a0cc845d86e7380c73269d2ee1d3c210/airflow/www/views.py#L3498)), but would like to check whether changing this behavior (which affect serialized DAG) is actually ok. Thanks!

bbovenzi replied:
The above is fine for now. But yes, I think we should show the conditions for triggering a DAG in the graph. That is separate from showing dataset aliases at all.

Lee-W replied:
@uranusjr I just created https://github.com/apache/airflow/pull/40984 to verify that deduplication is happening in the main branch and would like to confirm it's ok to change that behavior in this branch.

Lee-W replied:
@uranusjr I've changed the serialization behavior and moved the deduplication to the API layer.

uranusjr commented on airflow/serialization/dag_dependency.py:
Now that this is used for things other than serialisation, I wonder if the class should be moved outside of `serialization` entirely, maybe into `utils` or somewhere else.

Lee-W replied:
I feel it is still something mainly for `serialisation`. `iter_dag_deps` is mainly just for `serialisation` and later be used in the view.py. 

I do not object to moving it to `utils`. I'm just not sure moving lots of things there is a good idea 🤔

uranusjr replied:
Fair enough, it’s an internal class so most places are fine as long as contributors can find it and it doesn’t introduce import cycle issues.

Lee-W committed:
refactor(dataset): rename iter_dag_deps as iter_dag_dependencies

Lee-W committed:
refactor(dataset): move iter_dag_dependencies deduplicate out and do deduplicate in where it's called

uranusjr commented on airflow/datasets/__init__.py:
```diff
@@ -306,6 +350,9 @@ def __init__(self, name: str) -> None:
         self.name = name
         self.objects = expand_alias_to_datasets(name)
 
+    def __repr__(self) -> str:
+        return f"_DatasetAliasCondition({', '.join(map(str, self.objects))})"
```
This does not seem right? At least the class name certainly isn’t.

Lee-W replied:
This is actually expected. I'd like to use it to separate `DatasetAny` and `_DatasetAliasCondition`.

uranusjr commented on airflow/datasets/__init__.py:
```diff
@@ -314,6 +361,38 @@ def as_expression(self) -> Any:
+        """
+        if self.objects:
+            for obj in self.objects:
+                dataset = cast(Dataset, obj)
```
I don’t think this is right? We can have nesting conditions, in those cases `obj` would not be a Dataset.

Lee-W replied:
No nested condition would happen in `_DatasetAliasCondition` (at least in the current design) . This condition will only expand a `DatasetAlias` to `DatasetAny( resolved datasets )`

Lee-W committed:
refactor(serialization): remove dataset deduplication

Lee-W committed:
feat(serialization): make dag_deps lists instead of set

Lee-W committed:
feat(serialization): move the deduplication to the API layer instead of serialized DAG

(i.e., in serialzed dag, we'll have duplicate dag deps if users define this way)
